src/data/data.cc

```diff
@@ -29,7 +31,96 @@ DMLC_REGISTRY_ENABLE(::xgboost::data::SparsePageFormatReg<::xgboost::SortedCSCPa
 }  // namespace dmlc
 
+namespace {
+
+template <typename T>
+void SaveScalarField(dmlc::Stream *strm, const std::string &name,
+                     xgboost::DataType type, const T &field) {
+  strm->Write(name);
+  strm->Write(type);
+  strm->Write(true);  // is_scalar=True
+  strm->Write(field);
+}
+
+template <typename T>
+void SaveVectorField(dmlc::Stream *strm, const std::string &name,
+                     xgboost::DataType type, std::pair<uint64_t, uint64_t> shape,
+                     const std::vector<T>& field) {
+  strm->Write(name);
+  strm->Write(type);
+  strm->Write(false);  // is_scalar=False
+  strm->Write(shape.first);
+  strm->Write(shape.second);
+  strm->Write(field);
+}
+
+template <typename T>
+void SaveVectorField(dmlc::Stream* strm, const std::string& name,
+                     xgboost::DataType type, std::pair<uint64_t, uint64_t> shape,
+                     const xgboost::HostDeviceVector<T>& field) {
+  SaveVectorField(strm, name, type, shape, field.ConstHostVector());
+}
+
+template <typename T>
+void LoadScalarField(dmlc::Stream* strm, const std::string& expected_name,
+                     xgboost::DataType expected_type, T* field) {
+  const std::string invalid {"MetaInfo: Invalid format. "};
+  std::string name;
+  xgboost::DataType type;
+  bool is_scalar;
+  CHECK(strm->Read(&name)) << invalid;
+  CHECK_EQ(name, expected_name)
+      << invalid << " Expected field: " << expected_name << ", got: " << name;
+  CHECK(strm->Read(&type)) << invalid;
+  CHECK(type == expected_type)
+      << invalid << "Expected field of type: " << static_cast<int>(expected_type) << ", "
+      << "got field type: " << static_cast<int>(type);
+  CHECK(strm->Read(&is_scalar)) << invalid;
+  CHECK(is_scalar)
+    << invalid << "Expected field " << expected_name << " to be a scalar; got a vector";
+  CHECK(strm->Read(field, sizeof(T))) << invalid;
+}
+
+template <typename T>
+void LoadVectorField(dmlc::Stream* strm, const std::string& expected_name,
+                     xgboost::DataType expected_type, std::vector<T>* field) {
+  const std::string invalid {"MetaInfo: Invalid format. "};
+  std::string name;
+  xgboost::DataType type;
+  bool is_scalar;
+  CHECK(strm->Read(&name)) << invalid;
+  CHECK_EQ(name, expected_name)
+    << invalid << " Expected field: " << expected_name << ", got: " << name;
+  CHECK(strm->Read(&type)) << invalid;
+  CHECK(type == expected_type)
+    << invalid << "Expected field of type: " << static_cast<int>(expected_type) << ", "
+    << "got field type: " << static_cast<int>(type);
+  CHECK(strm->Read(&is_scalar)) << invalid;
+  CHECK(!is_scalar)
+    << invalid << "Expected field " << expected_name << " to be a vector; got a scalar";
+  std::pair<uint64_t, uint64_t> shape;
+
+  CHECK(strm->Read(&shape.first));
+  CHECK(strm->Read(&shape.second));
+  // TODO(hcho3): this restriction may be lifted, once we add a field with more than 1 column.
```

Comment:
@trivialfis Thanks for your suggestion to add a second shape dimension (column). This is going to be useful to express the label range in survival analysis.